File: Week_07/1091/shortest_path_in_binary_matrix.go
package main

type point struct {
	x int
	y int
}

func shortestPathBinaryMatrix(grid [][]int) int {
	row := len(grid)
	column := len(grid[0])
	if grid[0][0] == 1 || grid[row-1][column-1] == 1 {
		return -1
	}

	queue := []point{point{0, 0}}
	res := 0
	offsets := [8]point{
		{0, 1},
		{0, -1},
		{-1, -1},
		{-1, 0},
		{-1, 1},
		{1, 0},
		{1, -1},
		{1, 1},
	}
	for len(queue) > 0 {
		res++
		l := len(queue)
		for i := 0; i < l; i++ {
			x, y := queue[i].x, queue[i].y
			if x == row-1 && y == column-1 {
				return res
			}
			for _, o := range offsets {
				nextX, nextY := x+o.x, y+o.y
				if nextX >= 0 && nextX < row && nextY >= 0 && nextY < column && grid[nextX][nextY] == 0 {
					queue = append(queue, point{nextX, nextY})
					grid[nextX][nextY] = 1
				}
			}
		}
		queue = queue[l:]
	}
	return -1
}

File: Week_07/1091/shortest_path_in_binary_matrix.py
from typing import List


class Solution:
    def shortestPathBinaryMatrix(self, grid: List[List[int]]) -> int:
        m, n = len(grid) - 1, len(grid[0]) - 1
        if grid[0][0] == 1 or grid[m][n] == 1:
            return -1

        offsets = [
            (0, 1),
            (0, -1),
            (1, -1),
            (1, 0),
            (1, 1),
            (-1, -1),
            (-1, 0),
            (-1, 1),
        ]

        queue = [(0, 0)]
        res = 0
        while queue:
            res += 1
            l = len(queue)
            for i in range(l):
                if queue[i][0] == m and queue[i][1] == n:
                    return res
                for dx, dy in offsets:
                    x, y = queue[i][0] + dx, queue[i][1] + dy
                    if 0 <= x <= m and 0 <= y <= n and grid[x][y] == 0:
                        queue.append((x, y))
                        grid[x][y] = 1
            queue = queue[l:]
        return -1
